### DISCUSSION TITLE
Dynamic Column Widths for crosstab and tables based on highlights +styles

### DISCUSSION THREAD

**Discussion by drydenem** [eclipse-birt/birt#1402] [Q&A]
Hello,
I have a couple questions regarding dynamic column widths:
1. Is it possible to have different column widths in a crosstab? For example, can the last column be 4in when the others are 2in? Or even better, can the column widths adjust to the width of the text in the column?
2. Is it possible to set column width in a table or crosstab through a highlight +style? Example: I created a style called Column_width that sets width=2 in, then I create a highlight that says when string_value = 'Yes' apply style "Column_width". I am able to set backgrounds like this, but width never works.

I have also tried to set row height via highlight + style and that also does not work for me. Any insights would be great!

**Comment by speckyspooky** [comment]
The change of table columns & rows at their dimensions width & height is possible of standard tables.
The change of styles can be done based on dynamic changes of an existing style or directly on property level of the according objects.

But this won't be work on crosstabs. The crostabs are not a standard table structure.
It is a special object which looks like a table on design time but there isn't any objects like column and rows (it seems only to be).
The mechanis which I mentioned for the standard table doesn't work on crosstabs.

Attached I have a short example of the style changes & property changes for an standard table.
I hope it will help you a little bit.
From my point of view I doesn't use crosstabs because the usage isn't so easy and the technical support of this object-element is more limited like standard table.

[table-style-dynamic.zip](https://github.com/eclipse-birt/birt/files/12490022/table-style-dynamic.zip)

Here an screen example:

![2023-08-31 21_39_09-BIRT Report Viewer – Mozilla Firefox](https://github.com/eclipse-birt/birt/assets/41593722/51f10aa6-a1a2-482b-845d-698cb83f368a)

**Reply by hvbtup** [reply]
I agree with Thomas; I tend to avoid crosstabs. ForJDBC data, in many cases it is possible to use some advanced SQL and then use a table item in BIRT instead of a crosstab.

For tables, it is possible to dynamically change column widths. This can be done in the onCreate event of the table item.

However, it is not possible to adapt the column width based on the string width. You could use the maximum character count of the column`s data, but when using a proportional font, that doesn't help much, eg. WWWW is much wider then IIII.

**Comment by drydenem** [comment]
Thanks for this! I will try to implement the code you have for my tables-that should be helpful for a few issues I'm facing. It's unfortunate that crosstabs can't dynamically change column width. I love using them so that I can create reports that work for different types of data structures we have--a sort of one sql query fits all.
